src/store/modules/DownloadModule.ts

```diff
@@ -23,34 +29,65 @@ export const DownloadModule = {
+            const assignId = state.allDownloads.length;
+            const downloadObject: DownloadProgress = {
+                assignId,
+                initialMods,
+                modName: '',
+                progress: 0,
+                failed: false,
+            };
+            state.allDownloads = [...state.allDownloads, downloadObject];
+            return assignId;
+        },
+    },
+
     getters: <GetterTree<State, RootState>>{
         activeDownloadCount(state) {
-            const active = state.allVersions.filter(
-                dl => !dl[1].failed && dl[1].progress < 100
+            const active = state.allDownloads.filter(
+                dl => !dl.failed && dl.progress < 100
             );
             return active.length;
         },
+        currentDownload(state) {
+            return state.allDownloads[state.allDownloads.length-1] || null;
+        },
     },
 
     mutations: {
-        setDownloadObject(state: State, downloadObject: DownloadProgress) {
-            state.downloadObject = {...downloadObject};
-        },
-        pushDownloadObjectToAllVersions(state: State, params: { assignId: number, downloadObject: DownloadProgress }) {
-            state.allVersions = [...state.allVersions, [params.assignId, params.downloadObject]];
-        },
-        updateDownloadObject(state: State, params: { assignId: number, downloadVersion: [number, DownloadProgress]}) {
-            const newVersions = [...state.allVersions];
-            newVersions[params.assignId] = params.downloadVersion;
-            state.allVersions = newVersions;
+        updateDownload(state: State, update: UpdateObject) {
+            const newDownloads = [...state.allDownloads];
+            const index = newDownloads.findIndex((old: DownloadProgress) => {
+                return old.assignId === update.assignId;
+            });
+
+            if (index === -1) {
+                // The DownloadProgress by the ID from the update wasn't found at all.
+                const err = new R2Error(
+                    'Failed to update download status.',
+                    `DownloadProgress with assign id of ${update.assignId} wasn't found.`,
+                    'Try initiating the download again or restarting the app.'
+                );
+                // Only throw if the download failed, otherwise just console.log() it.
+                if (update.failed) {
+                    throw err;
+                }
+                console.warn(err);
+                return;
+            }
+
+            if (index !== update.assignId) {
+                // Just as a bonus, log if the DownloadProgress by the ID was located at the corresponding index of the array.
+                console.log(`There was a mismatch between download update\'s assign ID (${update.assignId}) and the index it was found at (${index}).`);
+            }
+
+            newDownloads[update.assignId] = {...newDownloads[update.assignId], ...update};
```

Comment:
Critical(?): isn't this a bug? Since we can't trust that `index === update.assignId`, shouldn't we use `index` to index the arrays here?